### PR TITLE
[FEATURE] Added Catalog Product Scope Hints

### PR DESCRIPTION
**Pull request by lewisvoncken**
### Description

When there is a value set (when `Use Default` has been disabled) on website or storeview level there will be shown a notice in the Product Form

### Manual testing scenarios

1. Set up a Multi Website
2. Create a Product through the Admin
3. Change the Scope to a storeview
4. Add a Storeview or website Specific Value
5. Change the Scope to the Default Scope
6. Now you can see that there is a Storeview or Website Specific value for the changed attribute
7. Click on the link and then it will open the Product Edit Page on that specific Storeview (when it is on Website scope it will open the default store of that website because it is not possible to change scope to website)

image

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

File: app/code/Magento/Catalog/Ui/DataProvider/Product/Form/Modifier/Eav.php [hunk 1]
```diff
@@ -614,6 +624,149 @@ private function isProductExists()
...
+    }
+
+    /**
+     * @param array $overridden
+     * @return string
+     * @throws \Magento\Framework\Exception\LocalizedException
+     */
+    private function formatOverriddenScopes(array $overridden)
+    {
+        $formatted = '<div class="overridden-hint-wrapper">' .
```
Comment (sidolov) [inline app/code/Magento/Catalog/Ui/DataProvider/Product/Form/Modifier/Eav.php]: Try to render HTML in the view layer (html file for the field). In the modifier, we are allowed to return only raw data

File: app/code/Magento/Catalog/Ui/DataProvider/Product/Form/Modifier/Eav.php [hunk 2]
```diff
@@ -260,7 +268,8 @@ public function __construct(
         $attributesToEliminate = [],
         CompositeConfigProcessor $wysiwygConfigProcessor = null,
         ScopeConfigInterface $scopeConfig = null,
-        AttributeCollectionFactory $attributeCollectionFactory = null
+        AttributeCollectionFactory $attributeCollectionFactory = null,
+        Url $urlBuilder
```
Comment (sidolov) [inline app/code/Magento/Catalog/Ui/DataProvider/Product/Form/Modifier/Eav.php]: Please, inject new dependencies according to our [Backward Compatible Development Guide](https://devdocs.magento.com/guides/v2.3/contributor-guide/backward-compatible-development/)

**Review by sidolov — CHANGES_REQUESTED** [review]

File: app/code/Magento/Catalog/Ui/DataProvider/Product/Form/Modifier/Eav.php (hunk 2 above)
Comment (lewisvoncken) [reply app/code/Magento/Catalog/Ui/DataProvider/Product/Form/Modifier/Eav.php]: @sidolov 
I have just resolved this. Thanks for the feedback

File: app/code/Magento/Catalog/Ui/DataProvider/Product/Form/Modifier/Eav.php (hunk 1 above)
Comment (lewisvoncken) [reply app/code/Magento/Catalog/Ui/DataProvider/Product/Form/Modifier/Eav.php]: @sidolov 
Thanks for the feedback this has been resolved.

**Comment by lewisvoncken** [comment]
@sidolov the requested changes have been made.

File: app/code/Magento/Ui/view/base/web/templates/form/field.html [hunk 3]
```diff
@@ -40,6 +40,18 @@
 
         <div class="admin__additional-info" if="$data.additionalInfo" html="$data.additionalInfo"></div>
 
+        <div class="admin__scope-hints" if="$data.scopeHints">
```
Comment (sidolov) [inline app/code/Magento/Ui/view/base/web/templates/form/field.html:43]: Please, create the new custom template especially for the product form, it's not a good idea to define `scopeHints ` for all form fields

**Review by sidolov — CHANGES_REQUESTED** [review]

File: app/code/Magento/Ui/view/base/web/templates/form/field.html (hunk 3 above)
Comment (lewisvoncken) [reply app/code/Magento/Ui/view/base/web/templates/form/field.html:43]: @sidolov 
It won't be shown when there are no scopeHints and I think that there has to be changed to much to only use a separate template but correct me if I am wrong. Besides that it is also easy extendable for the Category Edit

Comment (sivaschenko) [reply app/code/Magento/Ui/view/base/web/templates/form/field.html:43]: @lewisvoncken the problem is that your changes are introduced to `app/code/Magento/Ui/view/base/web/templates/form/field.html` that is used for all fields, not only product, but the PHP implementation is related only to catalog product. Please create a template that will be used only for catalog product form.

File: app/code/Magento/Catalog/Ui/DataProvider/Product/Form/Modifier/Eav.php [hunk 4]
```diff
@@ -614,6 +625,132 @@ private function isProductExists()
...
+    private function getOverriddenLevels(ProductAttributeInterface $attribute)
+    {
+        $overridden = [];
+        $tree = $this->getScopeTree();
+        if (!$tree) {
+            return $overridden;
+        }
+        $contextScopeId = $this->locator->getStore()->getId();
+
+        switch ($attribute->getScope()) {
```
Comment (sidolov) [inline app/code/Magento/Catalog/Ui/DataProvider/Product/Form/Modifier/Eav.php:641]: please, use if statement for current logic, it will be more readable

**Review by sidolov — CHANGES_REQUESTED** [review]

**Comment by sidolov** [comment]
Hi @lewisvoncken , you developed the nice feature, just a few changes are required! We are looking forward to merging your code to mainline!

**Comment by lewisvoncken** [comment]
Hi @sidolov 
I will try to provide fixes asap.

**Comment by sivaschenko** [comment]
@lewisvoncken closing this pull request due to inactivity. Feel free to reopen if you'd like to continue the progress on it.

**Pull request closed without merging**